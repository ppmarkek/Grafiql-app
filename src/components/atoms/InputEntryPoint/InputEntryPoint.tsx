import { Grid, Tabs } from '@mui/material';
import PlayArrowIcon from '@mui/icons-material/PlayArrow';
import ContentCopyIcon from '@mui/icons-material/ContentCopy';
import * as monaco from 'monaco-editor';
import MonacoEditor from '@monaco-editor/react';
import {
  ButtonGrid,
  ButtonStart,
  GridTabs,
  StyledButton,
  StyledTab,
  StyledTextField,
} from './style';
import { useState } from 'react';

interface TabPanelProps {
  children?: React.ReactNode;
  index: number;
  value: number;
}

function CustomTabPanel(props: TabPanelProps) {
  const { children, value, index, ...other } = props;

  return (
    <Grid
      role="tabpanel"
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
      height={'70%'}
    >
      {value === index && <Grid height={'100%'}>{children}</Grid>}
    </Grid>
  );
}

const InputEntryPoint = () => {
  const [value, setValue] = useState(0);
  const [inputValue, setInputValue] = useState('');

  const a11yProps = (index: number) => {
    return {
      id: `simple-tab-${index}`,
      'aria-controls': `simple-tabpanel-${index}`,
    };
  };

  const handleChange = (_event: React.SyntheticEvent, newValue: number) => {
    setValue(newValue);
  };

  const handleEditorChange = (value: string | undefined) => {
    setInputValue(value || '');
  };

  const options: monaco.editor.IStandaloneEditorConstructionOptions = {
    selectOnLineNumbers: true,
    roundedSelection: false,
    readOnly: false,
    cursorStyle: 'line',
    glyphMargin: true,
    folding: true,
    minimap: {
      enabled: false,
    },
    autoClosingBrackets: 'always',
    autoClosingQuotes: 'always',
    autoSurround: 'languageDefined',
    bracketPairColorization: { enabled: true },
  };

  const copyToClipboard = async () => {
    if (inputValue) {
      try {
        await navigator.clipboard.writeText(inputValue);
        console.log('Text copied to clipboard');
      } catch (err) {
        console.error('Failed to copy: ', err);
      }
    }
  };

  return (
    <Grid container item xs={6} height={'90%'} flexDirection={'column'}>
      <Grid container height={'75%'}>
        <Grid width={'93%'}>
          <MonacoEditor
            data-testid="monaco-editor"
            height="100%"
            theme="vs-dark"
            language="graphql"
            value={inputValue}
            onChange={handleEditorChange}
            options={options}
          />
        </Grid>
        <ButtonGrid container>
          <ButtonStart data-testid="play-button">
            <PlayArrowIcon />
          </ButtonStart>
          <StyledButton onClick={copyToClipboard} data-testid="copy-button">
            <ContentCopyIcon />
          </StyledButton>
        </ButtonGrid>
      </Grid>
      <GridTabs container>
        <Grid sx={{ borderBottom: 1, borderColor: 'divider' }}>
          <Tabs
            value={value}
            onChange={handleChange}
            aria-label="basic tabs example"
          >
            <StyledTab label="Variables" {...a11yProps(0)} />
            <StyledTab label="Headers" {...a11yProps(1)} />
          </Tabs>
        </Grid>
        <CustomTabPanel value={value} index={0}>
          <StyledTextField multiline variant="filled" />
        </CustomTabPanel>
        <CustomTabPanel value={value} index={1}>
          <StyledTextField multiline variant="filled" />
        </CustomTabPanel>
      </GridTabs>
    </Grid>
  );
};

export default InputEntryPoint;
