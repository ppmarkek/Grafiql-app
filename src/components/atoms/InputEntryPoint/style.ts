import { Button, Grid, Tab, TextField } from '@mui/material';
import { styled } from '@mui/system';

export const StyledTextField = styled(TextField)({
  height: '100%',
  width: '100%',
  '.MuiInputBase-root': {
    height: '100%',
  },
  textarea: {
    color: '#fff',
    height: '100% !important',
  },
});

export const ButtonGrid = styled(Grid)({
  width: '7%',
  background: 'rgba(0, 0, 0, 0.06)',
  justifyContent: 'flex-start',
  alignItems: 'center',
  flexDirection: 'column',
  paddingTop: '15px',
  gap: '15px',
});

export const ButtonStart = styled(Button)({
  all: 'unset',
  width: '40px',
  height: '40px',
  background: '#FF5794',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  borderRadius: '8px',
  cursor: 'pointer',
  border: '1px solid #FF5794',
  transition: 'all 0.5s',

  span: {
    display: 'none',
  },
});

export const StyledButton = styled(Button)({
  all: 'unset',
  width: '42px',
  height: '42px',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  borderRadius: '8px',
  cursor: 'pointer',
  transition: 'all 0.5s',

  span: {
    display: 'none',
  },
});

export const GridTabs = styled(Grid)({
  flexDirection: 'column',
  backgroundColor: 'rgba(0, 0, 0, 0.09)',
  borderBottomLeftRadius: '4px',
  borderBottomRightRadius: '4px',
  height: '25%',
});

export const StyledTab = styled(Tab)({
  color: '#fff',
});
