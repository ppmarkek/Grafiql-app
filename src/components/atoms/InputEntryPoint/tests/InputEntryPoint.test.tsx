import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import InputEntryPoint from '../InputEntryPoint';
import { Grid } from '@mui/material';

interface TabPanelProps {
  children?: React.ReactNode;
  index: number;
  value: number;
}

function CustomTabPanel(props: TabPanelProps) {
  const { children, value, index, ...other } = props;

  return (
    <Grid
      role="tabpanel"
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
      height={'70%'}
    >
      {value === index && <Grid height={'100%'}>{children}</Grid>}
    </Grid>
  );
}

jest.mock('@monaco-editor/react', () => {
  return ({
    onChange,
    value,
  }: {
    onChange: (value: string) => void;
    value: string;
  }) => (
    <textarea
      data-testid="mock-monaco-editor"
      onChange={(e) => onChange(e.target.value)}
      value={value}
    ></textarea>
  );
});

describe('InputEntryPoint Component', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(<InputEntryPoint />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles tab change', async () => {
    render(<InputEntryPoint />);
    const tab = screen.getByLabelText('Headers');
    await userEvent.click(tab);
  });

  it('handles input change in MonacoEditor', async () => {
    render(<InputEntryPoint />);
    const editor = await screen.findByRole('textbox');
    await userEvent.type(editor, 'new value');
  });

  it('triggers action when play button is clicked', async () => {
    render(<InputEntryPoint />);
    const playButton = await screen.findByTestId('play-button');
    await userEvent.click(playButton);
  });

  it('copies text to clipboard when copy button is clicked', async () => {
    render(<InputEntryPoint />);
    const copyButton = await screen.findByTestId('copy-button');
    await userEvent.click(copyButton);
  });

  it('displays correct tab panel on tab click', async () => {
    render(<InputEntryPoint />);
    const headersTab = screen.getByLabelText('Headers');
    const variablesTab = screen.getByLabelText('Variables');
    await userEvent.click(headersTab);
    await userEvent.click(variablesTab);
  });

  it('renders CustomTabPanel based on state value', () => {
    render(<InputEntryPoint />);
    const tabPanels = screen.getAllByRole('tabpanel');
    const variablesTabContent = tabPanels.find(
      (panel) => panel.getAttribute('id') === 'simple-tabpanel-0'
    );
    expect(variablesTabContent).toHaveAttribute('id', 'simple-tabpanel-0');
  });

  it('handles undefined input change in MonacoEditor', async () => {
    render(<InputEntryPoint />);
    const editor = await screen.findByTestId('mock-monaco-editor');
    fireEvent.change(editor, { target: { value: 'new value' } });
    expect((editor as HTMLTextAreaElement).value).toBe('new value');
  });

  it('handles error when copy to clipboard fails', async () => {
    Object.defineProperty(navigator, 'clipboard', {
      value: {
        writeText: jest.fn().mockRejectedValue(new Error('Failed to copy')),
      },
    });
    render(<InputEntryPoint />);
    const copyButton = await screen.findByTestId('copy-button');
    await userEvent.click(copyButton);
  });

  it('renders CustomTabPanel without children', () => {
    render(<CustomTabPanel index={0} value={0} />);
    const tabPanel = screen.getByRole('tabpanel');
    expect(tabPanel).toContainHTML('');
  });
});
