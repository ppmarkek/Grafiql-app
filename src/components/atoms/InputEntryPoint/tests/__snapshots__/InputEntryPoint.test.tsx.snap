// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputEntryPoint Component renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-6 css-kzse1c-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-10sxku-MuiGrid-root"
    >
      <div
        class="MuiGrid-root css-ajil6p-MuiGrid-root"
      >
        <textarea
          data-testid="mock-monaco-editor"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-12itivv-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1cg76yz-MuiButtonBase-root-MuiButton-root"
          data-testid="play-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PlayArrowIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8 5v14l11-7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-17h4eaq-MuiButtonBase-root-MuiButton-root"
          data-testid="copy-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ContentCopyIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-1fldk67-MuiGrid-root"
    >
      <div
        class="MuiGrid-root css-1dvh2m8-MuiGrid-root"
      >
        <div
          class="MuiTabs-root css-1ujnqem-MuiTabs-root"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              aria-label="basic tabs example"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="simple-tabpanel-0"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-q4f3tu-MuiButtonBase-root-MuiTab-root"
                id="simple-tab-0"
                role="tab"
                tabindex="0"
                type="button"
              >
                Variables
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="simple-tabpanel-1"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-q4f3tu-MuiButtonBase-root-MuiTab-root"
                id="simple-tab-1"
                role="tab"
                tabindex="-1"
                type="button"
              >
                Headers
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        aria-labelledby="simple-tab-0"
        class="MuiGrid-root css-12ni89x-MuiGrid-root"
        id="simple-tabpanel-0"
        role="tabpanel"
      >
        <div
          class="MuiGrid-root css-a3s15r-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1pe1dto-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-85zwa9-MuiInputBase-root-MuiFilledInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline css-7209ej-MuiInputBase-input-MuiFilledInput-input"
                id=":r0:"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline css-7209ej-MuiInputBase-input-MuiFilledInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        aria-labelledby="simple-tab-1"
        class="MuiGrid-root css-12ni89x-MuiGrid-root"
        hidden=""
        id="simple-tabpanel-1"
        role="tabpanel"
      />
    </div>
  </div>
</DocumentFragment>
`;
